### ISSUE TITLE
Filter out Marketo's email tracker from URLs

### ISSUE THREAD

**Issue by fmarier** [brave/brave-browser#9018]
The Marketo `mkt_tok` query string parameter is [for tracking users](https://nation.marketo.com/t5/Product-Discussions/How-to-remove-the-mkt-tok-query-parameter-added-to-the-end-of-my/m-p/139562/highlight/true#M95063) and [can be disabled by campaign admins](https://docs.marketo.com/display/public/DOCS/Disable+Tracking+for+an+Email+Link) though apparently it's [also used in unsubscribe links](https://nation.marketo.com/t5/Product-Discussions/What-s-the-use-of-mkt-tok-in-emails/m-p/147068/highlight/true#M101429).

Here is a longer description of how this parameter is used and what the security implications are: https://medium.com/@thezedwards/facebook-s-ongoing-vip-user-data-exfiltration-vulnerability-via-adobes-marketo-software-why-d8435a259b0

Need to find a real example and test the unsubscribe link before we can strip out in our query string filter.

**Comment by frankfuu** [comment]
Not sure if this is related but I'm also having issues editing a draft of an email when using Brave. It doesn't happen with Google Chrome.

![image](https://user-images.githubusercontent.com/7758528/81087854-8a0b6100-8f3d-11ea-94d8-311f5521bb3c.png)

**Comment by fmarier** [comment]
It's definitely not related to this issue since it hasn't been implemented yet :)

My guess is that it might be related to the referrer protections in Brave. Could you please file a new issue with more details on what tools you are using so that we can investigate?

**Comment by frankfuu** [comment]
Sweet, yep so it turns out it was related to those protections. Should I file a bug anyway or not worth?

**Comment by fmarier** [comment]
> Sweet, yep so it turns out it was related to those protections. Should I file a bug anyway or not worth?

Yes, please. We do try to address as many of the webcompat issues as we can.

**Comment by frankfuu** [comment]
Done - https://github.com/brave/brave-browser/issues/9665

**Comment by fmarier** [comment]
@frankfuu Do you happen to have a Marketo mailing list (test or real) I could join while testing the Brave filter?

I would like to make sure that when we block the user-tracking marketo query string parameter, we don't break unsubscribe links.

**Comment by frankfuu** [comment]
@fmarier I'm not sure what you mean by mailing list? But if it's access to our Marketo instance then I don't think my employer would agree. 

I'm happy to try and help test it though?

**Comment by fmarier** [comment]
By "mailing list" I meant some mailout list of some sort. Basically sending me an test email (to francois@brave.com) from within Marketo.

Ideally the body of the test email should include a link to say <https://brave.com>. What I would then like to do is click the unsubscribe link in that email to make sure that it works (i.e. you can see I unsubscribed).

**Comment by frankfuu** [comment]
No probs. I've sent you our newsletter email. At the bottom of the email there is an unsubscribe link which takes you to an email preferences center where you can untick the emails you do not want and when you press "save". Those changes will persist against your email. You can test that it has persisted by closing and re-opening the email and clicking on unsubscribe again.

**Comment by frankfuu** [comment]
Update: I just noticed you asked for a link in the email so I've resent you another email. It should be an email from f.fu@investsmart.com.au

**Comment by fmarier** [comment]
Thanks for the emails @frankfuu !

I can confirm that the `https://brave.com` link end up as: `https://brave.com/?mkt_tok=eyJpIjoiTURVd09HTmlNalEyTURneCIsInQiOiJcL2Z4TkxVOHRXUmNvcnFNaG1zS28zS2x6UWRDVEZqZE1Ma2dBcm1lNHBqdFJkbU1BRkI2V0gwNTNoMDRva2pmTEk1UTBOM2NqSXRhbVNIbXFJbUlhM0ZuY1dtbFZndFBEWUQrSEkxaHNYRXNGUGR2ZnBrOTgraWw1d29UaXNYaEYifQ%3D%3D` via a meta redirect:
```
$ curl -i https://go.investsmart.com.au/c0y01w3M09wW0fKGQ0pIz00
HTTP/2 200 
date: Thu, 21 May 2020 01:35:42 GMT
content-type: text/html
set-cookie: __cfduid=d1b43fd21fd0a808af2ac56f7681853851590024942; expires=Sat, 20-Jun-20 01:35:42 GMT; path=/; domain=.go.investsmart.com.au; HttpOnly; SameSite=Lax
cache-control: private, no-cache, no-store, max-age=0
x-content-type-options: nosniff
vary: Accept-Encoding
set-cookie: BIGipServersn_email_track_80=!03JRPFhwGTymMLQ+hMntxoNXdwmmDJ8o1AoAHceQePA6p5rhnodYjQwc46Kapocrr+yCqX8Vk1GzzC/Ewto3YQeSf+476EmBzUKrwDQ=; path=/; Httponly; Secure
cf-cache-status: DYNAMIC
expect-ct: max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"
server: cloudflare
cf-ray: 596a8fef8c023adc-YVR
cf-request-id: 02d67849b600003adcb1a47200000001

<html>
<head>
<meta charset='UTF-8'>
<meta name='robots' content='noindex'>
<script language='javascript'>
  var redirecturl = 'https://brave.com/?mkt_tok=eyJpIjoiTURVd09HTmlNalEyTURneCIsInQiOiJcL2Z4TkxVOHRXUmNvcnFNaG1zS28zS2x6UWRDVEZqZE1Ma2dBcm1lNHBqdFJkbU1BRkI2V0gwNTNoMDRva2pmTEk1UTBOM2NqSXRhbVNIbXFJbUlhM0ZuY1dtbFZndFBEWUQrSEkxaHNYRXNGUGR2ZnBrOTgraWw1d29UaXNYaEYifQ%3D%3D';
  function redirect() {
    var anchor = window.location.hash;
    window.self.location = redirecturl + anchor;
  }
</script>
</head>
<body onload=redirect()></body>
</html>
```
The `mkt_tok` tracking parameter decodes to the following JSON:
```
{
    "i": "MDUwOGNiMjQ2MDgx",
    "t": "/fxNLU8tWRcorqMhmsKo3KlzQdCTFjdMLkgArme4pjtRdmMAFB6WH053h04okjfLI5Q0N3cjItamSHmqImIa3FncWmlVgtPDYD+HI1hsXEsFPdvfpk98+il5woTisXhF"
}
```

The unsubscribe link (`https://pages.investsmart.com.au/UnsubscribePage.html?mkt_unsubscribe=1&mkt_tok=eyJpIjoiTURVd09HTmlNalEyTURneCIsInQiOiIvZnhOTFU4dFdSY29ycU1obXNLbzNLbHpRZENURmpkTUxrZ0FybWU0cGp0UmRtTUFGQjZXSDA1M2gwNG9ramZMSTVRME4zY2pJdGFtU0htcUltSWEzRm5jV21sVmd0UERZRCtISTFoc1hFc0ZQZHZmcGs5OCtpbDV3b1Rpc1hoRiJ9`) also includes a different `mkt_tok` tracking parameter, though it decodes same JSON structure.

It looks like it uses that parameter to prefill my email address in the page head:
```
<!DOCTYPE html>
<html lang="en"><head>
<script type="text/javascript">
 var mktoPreFillFields = {"Email":decodeURIComponent("francois%40brave.com"),"ER_Opt_in_A__c":true,"II_Opt_in_A__c":true,"IS_Opt_in_A__c":true};
</script>
```

**Comment by fmarier** [comment]
I tried to unsubscribe in two different ways:
1. From the "Hello world" email by visiting the page with the query string untouched and then reloading the page without the query string.
2. From the "Intelligent Investor" newsletter email by opening the link in curl and then opening `https://pages.investsmart.com.au/my-email-preferences?mkt_unsubscribe=1` (i.e. stripping out the `mkt_tok` parameter) directly in Brave.

@frankfuu Are you able to check whether I successfully unsubscribed from either or these lists or both?

**Comment by frankfuu** [comment]
Ah @fmarier we have a custom unsubscribe process so you actually have to uncheck each of the options individually. Not the best user experience I know but it was a business requirement. I can see your click and web visit activity from Marketo though. Not sure if that is enough to answer your question.

**Comment by fmarier** [comment]
@frankfuu Thanks, I think I have enough details to start thinking about what needs to be done here. It's not going to be as easy as with the other filters.

Thanks for your help in investigating this!

**Comment by frankfuu** [comment]
No problems @fmarier , let me know if you need more testing to be done ;)

**Comment by alfredonodo** [comment]
Any update on this?

**Comment by fmarier** [comment]
We've not yet found a way to remove this tracker without breaking unsubscribe pages.

The root of the problem is that Marketo's design uses the same identifier for tracking users as well as for authenticating them. Unless we can find a way to reliably detect and exempt unsubscribe pages, removing this tracker will not be as innocuous as the other ones we have removed so far.

**Comment by arthuredelstein** [comment]
After a [series of online searches](https://docs.google.com/document/d/1ZxI2y2vKDVW3A5qXQIFD0pLGVv33on2Fax2UVbUkYBs/), I am convinced that all (or nearly all) of unsubscribe links for Marketo emails contain the parameter `mkt_unsubscribe=1` in the query string. So my approach to fixing this bug is to filter out the `mkt_tok` query parameter EXCEPT where `mkt_unsubscribe=1` is also present in the query string.

**Comment by LaurenWags** [comment]
@arthuredelstein this issue is missing some key labels, can we get this reviewed please?

If manual QA is required let's get `QA/Yes` and a test plan added. If not, `QA/No` can be added.

Does this apply to Desktop, Android, or both? Please add the appropriate `OS/Desktop` and/or `OS/Android` labels.

Does this need to be included in the release notes or no? Please add the appropriate `release-notes/exclude` or `release-notes/include` label.

cc @fmarier @pes10k in case either of you have insight as well.

cc @brave/qa-team @rebron

**Comment by arthuredelstein** [comment]
Release note suggested by @pes10k : "improvements to Brave’s query-parameter stripping feature, to better protect users from cross-site tracking.”"

**Comment by LaurenWags** [comment]
excellent, thanks @arthuredelstein (and @pes10k 😄 )

**Comment by stephendonner** [comment]
9018-steps.md

## Verified `PASSED` using

Brave | 1.42.61 Chromium: 103.0.5060.114 (Official Build) beta (x86_64)
--|--
Revision | a1c2360c5b02a6d4d6ab33796ad8a268a6128226-refs/branch-heads/5060@{#1124}
OS | macOS Version 13.0 (Build 22A5295i)

### Shared steps:
1. install `1.42.61`
2. launch Brave
3. load each of the following URLs
4. confirm the respective parameters are handled appropriately, per each testcase

### Remove `mkt_tok`, general cases

* https://example.com/?mkt_tok=abc
* https://example.com/?mkt_tok=abc&another=test
* https://example.com/?param1=foo&mkt_tok=abc&another=test

### Retain `mkt_tok`

* https://example.com/unsubscribe.html?mkt_tok=abc
* https://example.com/?mkt_tok=abc&mkt_unsubscribe=1
* https://example.com/?param1=foo&mkt_tok=abc&unsubscribe=yes

`?mkt_tok=abc` | `?mkt_tok=abc&mkt_unsubscribe=1` | `?param1=foo&mkt_tok=abc&unsubscribe=yes`
-----------------|---------------------------------------|-------------------------------------------------



### Regression-test existing forbidden parameters
* https://example.com/?fbclid=1
* https://example.com/?gclid=1
* https://example.com/?msclkid=1

`?fbclid=1` | `?gclid=1` | `?msclkid=1`
-----------|-------------|-------------




### Query-filter tests

* https://fmarier.github.io/brave-testing/query-filter.html

**Comment by arthuredelstein** [comment]
Thank you @stephendonner!

**Comment by Uni-verse** [comment]
Verification `PASSED` on `Samsung Galaxy S21` using 

```
Brave	1.42.78 Chromium: 103.0.5060.134 (Official Build) beta (64-bit) 
Revision	8ec6fce403b3feb0869b0732eda8bd95011d333c-refs/branch-heads/5060@{#1262}
OS	Android 12; Build/SP1A.210812.016
```

**Removing mkt_tok**

?mkt_tok=abc | ?mkt_tok=abc&another=test | ?param1=foo&mkt_tok=abc&another=test
-|-|-
![screenshot-1658877187837](https://user-images.githubusercontent.com/17885425/181129638-78842f25-edbb-4ffb-820b-9ffcc789eb38.png) | ![screenshot-1658877203492](https://user-images.githubusercontent.com/17885425/181129644-9dd20702-d714-4a77-b5c4-d0738f8d519d.png) | ![screenshot-1658877215567](https://user-images.githubusercontent.com/17885425/181129645-f340c418-36f8-40f2-a1b4-fb0801b25eeb.png)


**Keeping the mkt_tok**

unsubscribe.html?mkt_tok=abc | ?mkt_tok=abc&mkt_unsubscribe=1 | ?param1=foo&mkt_tok=abc&unsubscribe=yes
-|-|- 
![screenshot-1658877133353](https://user-images.githubusercontent.com/17885425/181129727-e47cd472-8c18-45f4-a39f-4941a3d7e072.png) | ![screenshot-1658877152311](https://user-images.githubusercontent.com/17885425/181129730-6c0b89e8-d28c-4012-be49-7221ddcc2426.png) | ![screenshot-1658877167529](https://user-images.githubusercontent.com/17885425/181129731-2ebed750-bcbd-45f5-a53a-7ffb34f23bf2.png)


**Existing forbidden parameters**

- [x] https://example.com/?fbclid=1
- [x] https://example.com/?gclid=1
- [x] https://example.com/?msclkid=1


- [x] https://fmarier.github.io/brave-testing/query-filter.html

**Comment by Uni-verse** [comment]
Verification in Progress on Android ARM tablet using 

Brave	1.42.80 Chromium: 104.0.5112.57 (Official Build) (64-bit) 
Revision	212fd173a0da1e0a024f328295bb56a2529190bb-refs/branch-heads/5112@{#1042}
OS	Android 12; Build/SP1A.210812.016

**Comment by eligrey** [comment]
What's to stop Marketo from simply including `unsubscribe` in every URL to break this filter's usefulness? This seems feels very fragile.

I think that this filter should check for a `mkt_unsubscribe=1` query parameter specifically, and not use a whole URL substring search as indicated in the docs.
